pkg/services/queryhistory/api.go

```diff
@@ -15,15 +16,27 @@ import (
 
 func (s *QueryHistoryService) registerAPIEndpoints() {
 	s.RouteRegister.Group("/api/query-history", func(entities routing.RouteRegister) {
-		entities.Post("/", middleware.ReqSignedIn, routing.Wrap(s.createHandler))
-		entities.Get("/", middleware.ReqSignedIn, routing.Wrap(s.searchHandler))
-		entities.Delete("/:uid", middleware.ReqSignedIn, routing.Wrap(s.deleteHandler))
-		entities.Post("/star/:uid", middleware.ReqSignedIn, routing.Wrap(s.starHandler))
-		entities.Delete("/star/:uid", middleware.ReqSignedIn, routing.Wrap(s.unstarHandler))
+		entities.Post("/", middleware.ReqSignedIn, routing.Wrap(s.permissionsMiddleware(s.createHandler, "Failed to create query history")))
+		entities.Get("/", middleware.ReqSignedIn, routing.Wrap(s.permissionsMiddleware(s.searchHandler, "Failed to get query history")))
+		entities.Delete("/:uid", middleware.ReqSignedIn, routing.Wrap(s.permissionsMiddleware(s.deleteHandler, "Failed to delete query history")))
+		entities.Post("/star/:uid", middleware.ReqSignedIn, routing.Wrap(s.permissionsMiddleware(s.starHandler, "Failed to star query history")))
+		entities.Delete("/star/:uid", middleware.ReqSignedIn, routing.Wrap(s.permissionsMiddleware(s.unstarHandler, "Failed to unstar query history")))
 		entities.Patch("/:uid", middleware.ReqSignedIn, routing.Wrap(s.patchCommentHandler))
 	})
 }
 
+type CallbackHandler func(c *contextmodel.ReqContext) response.Response
+
+func (s *QueryHistoryService) permissionsMiddleware(next CallbackHandler, errorMessage string) CallbackHandler {
```

Comment:
@grafana/explore-squad let me know what you think about this approach. If you're ok with it, I can update other tests in the API to use this middleware. The approach would be similar to what I've done in queryhistory_create_test.go